// twilio-sync/browser/lql/helpers/syntaxer.js
"use strict";

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getLowPriorityBlock(tree, priority) {
    var lexemes = require('../config/lexemes');
    if (priority === undefined) {
        priority = -1;
    }
    if (!tree.lexeme || lexemes[tree.lexeme.type].priority < priority || tree.right === null) {
        if (tree.left && (0, _keys2.default)(tree.left).length == 0) {
            return tree.left;
        }
        return tree;
    }
    return getLowPriorityBlock(tree.right, priority);
}
function genericFieldSyntaxer(type) {
    return function (options, lexemesArray, i, tree) {
        var lexemes = require('../config/lexemes');
        var block = getLowPriorityBlock(tree, lexemes.string.priority);
        var tempLexeme = {
            type: type,
            value: lexemesArray[i].value
        };
        block.lexeme = tempLexeme;
        block.left = null;
        block.right = null;
        return i + 1;
    };
}
function generateBasicSyntaxer(type) {
    return function basicSyntaxer(options, lexemesArray, i, tree) {
        var lexemes = require('../config/lexemes');
        var block = getLowPriorityBlock(tree, lexemes[type].priority);
        if (!block.lexeme) {
            throw new Error('Nothing before ' + type + ' block');
        }
        var tempLexeme = {
            type: block.lexeme.type,
            value: block.lexeme.value
        };
        block.lexeme = {
            type: type
        };
        block.left = {
            lexeme: tempLexeme,
            left: block.left,
            right: block.right
        };
        block.right = {};
        if (!block.left.lexeme.value) {
            delete block.left.lexeme.value;
        }
        return i + 1;
    };
}
function notSyntaxer(options, lexemesArray, i, tree) {
    var lexemes = require('../config/lexemes');
    var block = getLowPriorityBlock(tree, lexemes['not'].priority);
    block.lexeme = {
        type: 'not'
    };
    block.left = {};
    block.right = null;
    return i + 1;
}
function genericBlockSyntaxer(typeStart, typeEnd) {
    return function basicSyntaxer(options, lexemesArray, i, tree, createTree) {
        var lexemes = require('../config/lexemes');
        var block = getLowPriorityBlock(tree, eval('lexemes.' + typeStart + '.priority'));
        var tempLexeme = {
            type: typeStart
        };
        var newBlock = createTree(typeEnd, i + 1);
        block.lexeme = tempLexeme;
        block.left = newBlock;
        block.right = null;
    };
}
module.exports = {
    andSyntaxer: generateBasicSyntaxer('and'),
    orSyntaxer: generateBasicSyntaxer('or'),
    eqSyntaxer: generateBasicSyntaxer('eq'),
    notEqualSyntaxer: generateBasicSyntaxer('not_eq'),
    commaSyntaxer: generateBasicSyntaxer('comma'),
    inSyntaxer: generateBasicSyntaxer('in'),
    notInSyntaxer: generateBasicSyntaxer('not_in'),
    containsSyntaxer: generateBasicSyntaxer('contains'),
    blockSyntaxer: genericBlockSyntaxer('startBlock', 'endBlock'),
    arraySyntaxer: genericBlockSyntaxer('startArray', 'endArray'),
    stringSyntaxer: genericFieldSyntaxer('string'),
    attrSyntaxer: genericFieldSyntaxer('attr'),
    notSyntaxer: notSyntaxer
};
